BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,199 @@
+        setContentView(binding.root)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        cryptographyManager = CryptographyManager()
+        val ciphertextWrapper = cryptographyManager.getCiphertextWrapperFromSharedPrefs(
+            applicationContext,
+            SHARED_PREFS_FILENAME,
+            Context.MODE_PRIVATE,
+            CIPHERTEXT_WRAPPER
+        )
+
+        if (ciphertextWrapper != null) {
+            if (SampleAppUser.fakeToken == null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                // The user has already logged in, so proceed to the rest of the app
+                // this is a todo for you, the developer
+                updateApp(getString(R.string.already_signedin))
+            }
+        } else {
+            loginWithPasswordManager()
+        }
+
+        val useBiometricsView = findViewById<AppCompatTextView>(R.id.use_biometrics)
+        useBiometricsView.setOnClickListener {
+            if (ciphertextWrapper != null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                startActivity(Intent(this, EnableBiometricLoginActivity::class.java))
+            }
+        }
+    }
+
+
+    // BIOMETRICS SECTION
+
+    private fun showBiometricPromptForDecryption() {
+        val canAuthenticate = BiometricManager.from(applicationContext).canAuthenticate()
+        if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
+            val secretKeyName = getString(R.string.secret_key_name)
+            val cipher = cryptographyManager.getInitializedCipherForDecryption(
+                secretKeyName, encryptedServerTokenWrapper.initializationVector
+            )
+            biometricPrompt =
+                BiometricPromptUtils.createBiometricPrompt(this, ::decryptServerTokenFromStorage)
+            val promptInfo = BiometricPromptUtils.createPromptInfo(this)
+            biometricPrompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
+        }
+    }
+
+    private fun decryptServerTokenFromStorage(authResult: BiometricPrompt.AuthenticationResult) {
+        authResult?.cryptoObject?.cipher?.apply {
+            val plaintext =
+                cryptographyManager.decryptData(encryptedServerTokenWrapper.ciphertext, this)
+            SampleAppUser.fakeToken = plaintext
+            // now that you have the token, you can query server for everything else
+            // the only reason we call this fakeToken is because we didn't really get it from
+            // the server. In your case, you will have gotten it from the server the first time
+            // and therefore, it's a real token.
+
+            updateApp(getString(R.string.already_signedin))
+        }
+    }
+
+    // USERNAME + PASSWORD SECTION
+
+    private fun loginWithPasswordManager() {
```

Comment:
Everything in this method looks like it should be done in onCreate?

The ViewModel should be created there. Views should be observed from there. etc...

Reply:
I find this modular approach easier to follow

Reply:
You can keep it modular and still call it from `onCreate`.

`loginWithPasswordManager` doesn't seem like the right name, since this function _doesn't_ log the user in, as far as I can tell. It sets up a bunch of listeners and observers, which should be setup in `onCreate`.

Reply:
ditto. 
Please take a moment to run the code in your IDE. If you move the logic from onResume to onCreate, you will introduce a bug in the flow.

Reply:
@nic0lette's naming comment is still valid, @isaidamier.

Reply:
Please suggest another name. As you can see there are two main user journeys in this Activity: login with biometrics or login with password. I group the "with password" section so devs/readers can more easily identify the biometrics portion.

Reply:
`bindViews` or `setUpInterface` could work.